thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java

```diff
@@ -266,36 +258,25 @@ public static THttpService ofFormats(
     public static Function<? super RpcService, THttpService> newDecorator(
             SerializationFormat defaultSerializationFormat,
             Iterable<SerializationFormat> otherSupportedSerializationFormats) {
-
-        final SerializationFormat[] supportedSerializationFormatArray = newSupportedSerializationFormats(
-                defaultSerializationFormat, otherSupportedSerializationFormats);
-
-        return delegate -> new THttpService(delegate, supportedSerializationFormatArray);
-    }
-
-    private static SerializationFormat[] newSupportedSerializationFormats(
-            SerializationFormat defaultSerializationFormat,
-            Iterable<SerializationFormat> otherSupportedSerializationFormats) {
-
-        requireNonNull(defaultSerializationFormat, "defaultSerializationFormat");
-        requireNonNull(otherSupportedSerializationFormats, "otherSupportedSerializationFormats");
-
-        final Set<SerializationFormat> set = new LinkedHashSet<>();
-        set.add(defaultSerializationFormat);
-        Iterables.addAll(set, otherSupportedSerializationFormats);
-        return set.toArray(EMPTY_FORMATS);
+        return builder().defaultSerializationFormat(defaultSerializationFormat)
+                        .otherSerializationFormats(otherSupportedSerializationFormats)
+                        .newDecorator();
     }
 
-    private final SerializationFormat[] supportedSerializationFormatArray;
-    private final Set<SerializationFormat> supportedSerializationFormats;
     private final ThriftCallService thriftService;
+    private final SerializationFormat defaultSerializationFormat;
+    private final Set<SerializationFormat> supportedSerializationFormats;
+    private final BiFunction<ServiceRequestContext, ? super Throwable, ? extends Throwable> exceptionMapper;
 
-    THttpService(RpcService delegate, SerializationFormat[] supportedSerializationFormatArray) {
+    THttpService(RpcService delegate,
+                 SerializationFormat defaultSerializationFormat,
+                 Set<SerializationFormat> supportedSerializationFormats,
+                 BiFunction<ServiceRequestContext, ? super Throwable, ? extends Throwable> exceptionMapper) {
```

Comment:
How about generalizing this more like `BiFunction<? super ServiceRequestContext, ? super Throwable, ? extends RpcResponse> exceptionHandler`?
I think we can use a similar approach we did in `GrpcServiceBuilder`
https://github.com/line/armeria/blob/480321970111ac6249775199fedadcc71b862e84/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java#L483-L500

Reply:
If we want to return `RpcResponse`, we have to add a decorator like we did in here. https://github.com/line/centraldogma/blob/master/server/src/main/java/com/linecorp/centraldogma/server/internal/thrift/CentralDogmaExceptionTranslator.java
So we can add a callback to propagate the event to here:
https://github.com/line/armeria/blob/master/thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java#L539
I'm not sure if it's worth.

Reply:
Thanks, changed to return `RpcResponse`. 😉